File: 211-design-add-and-search-words-data-structure/design-add-and-search-words-data-structure.py
class Node : 


    def __init__(self,):

        self.dic ={ }
        self.isEnd  = False

class WordDictionary:

    def __init__(self):
        
        self.head = Node()
        

    def addWord(self, word: str) -> None:

        temp =self.head

        for c in word:
            if c not in temp.dic:
                temp.dic[c] = Node()
            temp  = temp.dic[c]
        temp.isEnd =True
     
        

    def search(self, word: str) -> bool:
        n =len(word)

        def dfs(node, ind):
            if  ind ==n:
                return node.isEnd
            ch = word[ind]
            if ch==".":

                for c in node.dic:
                    if  dfs(node.dic[c],ind+1):
                        return True
                return False
            
            else:
                if ch not in node.dic:
                    return False
                return dfs(node.dic[ch],ind+1)
        return dfs(self.head, 0)

        
        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: 211-design-add-and-search-words-data-structure/design-add-and-search-words-data-structure.js

function Node(){

    this.dic = { };
    this.isEnd = false;
}


var WordDictionary = function() {

        this.head = new Node()
        
    
};

/** 
 * @param {string} word
 * @return {void}
 */
WordDictionary.prototype.addWord = function(word) {

    let temp = this.head;
    

    for (let c of word) {
            if (!(c in temp.dic)) {
                temp.dic[c] = new Node();
            }
            temp = temp.dic[c];
        }
    temp.isEnd = true;
};

/** 
 * @param {string} word
 * @return {boolean}
 */
WordDictionary.prototype.search = function(word) {
    let n = word.length;
   

    let dfs = (node, ind) =>{
        if (ind === n) {
            return node.isEnd;
        }
        let ch = word[ind];
        if (ch === ".") {
            for (let c in node.dic) {
                if (dfs(node.dic[c], ind + 1)) {
                    return true;
                }
            }
            return false;
        } else {
            if (!(ch in node.dic)) {
                return false;
            }
            return dfs(node.dic[ch], ind + 1);
        }
    }

    return dfs(this.head, 0);
};

/** 
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */